### PR TITLE
[google_maps_flutter] Android: only destroy mapView once

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

In https://github.com/flutter/plugins/pull/2704, we removed the mapView.onDestroy call because I thought that alone caused the crash in https://github.com/flutter/flutter/issues/55406.
It turned out the crash was due to the mapView.onDestroy gets called twice because activity's onDestory also triggers the `mapView.onDestroy`. And the second call is trying to destroy an already destroyed map view which caused a NPE. 
This PR fixes that with refactoring all the `mapView.onDestroy` to a single function that makes sure it is only called once.
Since we brings back the `mapView.onDestroy` in `dispose`, the memory leak caused crashed in https://github.com/flutter/flutter/issues/56442 is also fixed.

## Related Issues

https://github.com/flutter/flutter/issues/56442
https://github.com/flutter/flutter/issues/55406

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java [hunk 1]
```diff
@@ -891,6 +893,14 @@ private Application getApplication() {
     }
   }
 
+  private void destroyMapViewIfNecessary() {
+    if (mapViewDestroyed || mapView == null) {
+      return;
+    }
+    mapView.onDestroy();
+    mapViewDestroyed = true;
```
Comment (blasten) [inline packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:901]: what about just making `mapView = null;` ?

**Comment by blasten** [comment]
is it possible to add an e3e test?

File: packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java (hunk 1 above)
Comment (xster) [reply packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:901]: ditto, can't the state tracking be done implicitly with the mapView instance variable?

Comment (cyanglaz) [reply packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:901]: The NPE didn't happen on the mapView object, it is rather something inside map SDK had a NPE. 
By looking at the log: 
```
Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void com.google.maps.api.android.lib6.impl.bp.q()' on a null object reference
```
Probably the first mapView.onDestroy cleaned up some google map sdk internal resources and the second call tries the clean up the same resource again and got a NPE.
I also looked up the documentation and I don't find anything that indicates if MapView was destroyed.

**Comment by cyanglaz** [comment]
> is it possible to add an e2e test?

Not sure how test e2e test on app exit tho. Maybe we can set up a Robolectric test in the plugin tho? I can look into that.

File: packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java (hunk 1 above)
Comment (xster) [reply packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:901]: ya, I got that part. What we're saying is now that we made mapView unusable by calling onDestroy on it, let's just remove its reference. I don't see that done anywhere in the GoogleMapController. If you did, you wouldn't have to use a separate bool to track whether mapView is destroyed.

Comment (cyanglaz) [reply packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:901]: Ah I see. I was considering it but I wasn't sure because it'll remove the `final` on mapView.

Comment (xster) [reply packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:901]: sure, you'd have to edit the doc for getView() too. But maybe that's a feature rather than a 'bug' since it's weird to start with that you could have a public API to 'get' a destroyed view.

Comment (cyanglaz) [reply packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:901]: Sounds good. Will update that! thanks.

**Comment by blasten** [comment]
Based on https://github.com/flutter/flutter/issues/56442, it looks like the sequence mount/unmount/mount widget should trigger the bug?

**Comment by cyanglaz** [comment]
> Based on [flutter/flutter#56442](https://github.com/flutter/flutter/issues/56442), it looks like the sequence mount/unmount/mount widget should trigger the bug?

The bug is triggered due to a memory leak, we'd have to either 
1. check if the particular google map sdk internal object is leaked, or 
2. we run the same tests for a very long time to make sure it doesn't crash.
Im not sure how we can do 1. 
For 2, i'm not sure what is the definitive number of times to run, since it is based on devices and stuff. I'll give option 2 a try and see how it goes.

**Comment by cyanglaz** [comment]
@xster @blasten Added some tests to test if the map view is set to null. We can't really test if the mapView is "Destroyed" as there is no indicator from the map sdk to do so.

We also can't really load testing the memory leak, because the leaked resource seems to be related to the google map service, which requires we set up the api key in the example app.

We can't test the double "onDestroy" calls crash, as the it doesn't seem to crash in test environment. I think it might be a synchronized call that the crashes happened on a different thread.

File: packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java [hunk 2]
```diff
@@ -71,7 +71,7 @@
   private final AtomicInteger activityState;
   private final MethodChannel methodChannel;
   private final GoogleMapOptions options;
-  private final MapView mapView;
+  private MapView mapView;
```
Comment (blasten) [inline packages/google_maps_flutter/google_maps_flutter/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:74]: nit: needs `@Nullable` annotation.

File: packages/google_maps_flutter/google_maps_flutter/example/android/app/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 3]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (blasten) [inline packages/google_maps_flutter/google_maps_flutter/example/android/app/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: is this file required?

Comment (cyanglaz) [reply packages/google_maps_flutter/google_maps_flutter/example/android/app/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: Yes, it is required mocking `GoogleMap` as it is a final class.

**Comment by cyanglaz** [comment]
The tests are failing on stable:
```
java.lang.UnsatisfiedLinkError: no flutter in java.library.path

	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at io.flutter.embedding.engine.loader.FlutterLoader.startInitialization(FlutterLoader.java:120)
	at io.flutter.embedding.engine.loader.FlutterLoader.startInitialization(FlutterLoader.java:88)
	at io.flutter.view.FlutterMain.startInitialization(FlutterMain.java:48)
	at io.flutter.app.FlutterApplication.onCreate(FlutterApplication.java:24)
	at org.robolectric.android.internal.AndroidTestEnvironment.lambda$installAndCreateApplication$0(AndroidTestEnvironment.java:276)
	at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:75)
	at org.robolectric.android.internal.AndroidTestEnvironment.installAndCreateApplication(AndroidTestEnvironment.java:276)
	at org.robolectric.android.internal.AndroidTestEnvironment.setUpApplicationState(AndroidTestEnvironment.java:169)
	at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:301)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:243)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
@xster It seems like it cannot find flutter in path. Do you know what might have happened? Works on master tho

**Comment by cyanglaz** [comment]
I saw 
https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/view/FlutterMain.java#L143-L157
says we can't use roboletric

**Comment by xster** [comment]
option 1, don't use FlutterApplication in the example android manifest. Shouldn't be necessary for v2 embedding. 

option 2, just use that static call you linked to make it not actually load the native libraries (which it obviously can't during test) in the test

**Comment by cyanglaz** [comment]
I will go with option 2. We want the example app to run on V2 embedding if possible.
As the `setIsRunningInRobolectricTest` is deprecated, this would be a temporary solution. We would need eventually migrate out from using this method.
